### PR TITLE
Reduce usages of repackaged ASM 5

### PR DESCRIPTION
**Pull request by basil**
A more conservative version of #5524 that allows for a more gentle transition period for users of SCM API (i.e., most Jenkins users). Like that PR, we still pull in the version of Stapler that drops the dependency on the repackaged ASM 5 and also stop consuming the repackaged ASM 5 from Jenkins core. But unlike #5524, we continue to bundle the repackaged ASM 5 as a direct dependency for the benefit of users of older versions of SCM API. This allows for a smoother (albeit more prolonged) transition period. Once jenkinsci/scm-api-plugin#88 is widely adopted, the transition can be completed by removing the repackaged ASM 5 from core. Unlike #5524, I have not written changelog or upgrade entries for this change because there should be no visible impact to users.

### Proposed changelog entries

N/A

### Proposed upgrade guidelines

N/A

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by basil** [comment]
Note that unlike #5524, this does _not_ need to be placed in "on-hold" status because it is not blocked on the release of any other component.

**Review by timja — APPROVED** [review]
no reason to delay this i guess

File: core/pom.xml
Comment (jglick) [inline core/pom.xml:125]:
```suggestion
      For compatibility only; not included into the BOM for the same reason. TODO once
```

File: core/src/main/java/hudson/util/SubClassGenerator.java
Comment (jglick) [inline core/src/main/java/hudson/util/SubClassGenerator.java:41]: As in the earlier PR, I think this thing is unused. We should strive to eliminate uses of ASM wherever possible.

**Review by jglick — APPROVED** [review]
Forgot to approve.

File: core/src/main/java/hudson/util/SubClassGenerator.java
Comment (basil) [reply core/src/main/java/hudson/util/SubClassGenerator.java:41]: Agreed that this class should be removed and verified that there are no references with `usage-in-plugins`. Since this is supposed to be the conservative PR I'll mark it with `@Deprecated` in this PR and save the removal for the more aggressive PR (#5524).

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Review by res0nance — APPROVED** [review]

**Pull request merged**
